python/cuml/experimental/cluster/hdbscan.pyx

```diff
@@ -0,0 +1,550 @@
+            return None
+
+        if self._condensed_tree is None:
+            raw_tree = np.recarray(shape=(self.condensed_parent_.shape[0],),
+                                   formats=[np.intp, np.intp, float, np.intp],
+                                   names=('parent', 'child', 'lambda_val',
+                                          'child_size'))
+            raw_tree['parent'] = self.condensed_parent_
+            raw_tree['child'] = self.condensed_child_
+            raw_tree['lambda_val'] = self.condensed_lambdas_
+            raw_tree['child_size'] = self.condensed_sizes_
+
+        try:
+            from hdbscan.plots import CondensedTree
+        except Exception as e:
+            raise ImportError("hdbscan must be installed to use plots")
+
+        return CondensedTree(raw_tree,
+                             self.cluster_selection_epsilon,
+                             self.allow_single_cluster)
+
+    def _build_single_linkage_tree(self):
+
+        if not self.fit_called_:
+            return None
+
+        if self._single_linkage_tree is None:
+
+            with cuml.using_output_type("numpy"):
+                raw_tree = np.column_stack(
+                    (self.children_[0, :self.n_leaves_-1],
+                     self.children_[1, :self.n_leaves_-1],
+                     self.lambdas_[:self.n_leaves_-1],
+                     self.sizes_[:self.n_leaves_-1]))
+
+            raw_tree = raw_tree.astype(np.float64)
+
+        try:
+            from hdbscan.plots import SingleLinkageTree
+        except Exception as e:
+            raise ImportError("hdbscan must be installed "
+                              "to use plots")
+
+        return SingleLinkageTree(raw_tree)
+
+    def _build_minimum_spanning_tree(self, X):
+
+        with cuml.using_output_type("numpy"):
+            raw_tree = np.column_stack((self.mst_src_,
+                                        self.mst_dst_,
+                                        self.mst_weights_))
+
+        raw_tree = raw_tree.astype(np.float64)
+
+        try:
+            from hdbscan.plots import MinimumSpanningTree
+        except Exception as e:
+            raise ImportError("hdbscan must be installed to use plots")
+
+        self.minimum_spanning_tree_ = MinimumSpanningTree(
+            raw_tree, X.to_output("numpy"))
+
+    def __dealloc__(self):
+        delete_hdbscan_output(self)
+
+    def _cuml_array_from_ptr(self, ptr, buf_size, shape, dtype):
+
+        mem = cp.cuda.UnownedMemory(ptr=ptr, size=buf_size,
+                                    owner=self.hdbscan_output_,
+                                    device_id=-1)
+        mem_ptr = cp.cuda.memory.MemoryPointer(mem, 0)
+
+        return CumlArray(data=cp.ndarray(shape=shape,
+                                         dtype=dtype,
+                                         memptr=mem_ptr)).to_output('numpy')
+
+    def _construct_condensed_tree_attribute(self, ptr, dtype="int32"):
```

Comment:
Yeah, they are just name choices. I think mentally I put "construct" with things which might contain less trivial code (such as manipulating cython pointers) and "build" with things that are just a grouping of higher-level API calls. I can make them consistent if that would be more readable.

On a side note- I had originally wanted to do what is being in done in HDBSCAN where the condensed_tree_ attribute was a property which could be constructed lazily but it was throwing some strange errors in the base class (because hasattr is already being hijacked there). I wouldn't mind trying this again if you have any suggestions about whether I'm doing it wrong or how I might be able to overcome the issue. I tried using `@property`

Reply:
Can you add the item of trying this as a property, I think it can be a follow up nice to have but not indispensable (and need to think about the getattr…)